### ISSUE TITLE
redundant_supertrait_bound: warn when trait bound is already required by supertrait

### ISSUE THREAD

**Issue by oriongonza** [rust-lang/rust-clippy#16162]
### What it does

```rs
trait Trait: Send {}
fn foo<T: Trait + Send>() {}
//                ^^^^ redundant
```

`Send` is already required by `Trait`. The explicit bound adds noise without changing semantics.

This lint should either be `complexity` or `pedantic` — it's not wrong, just unnecessary.

**Proposed Message**
```
warning: redundant trait bound
  --> src/main.rs:2:18
   |
2  | fn foo<T: Trait + Send>() {}
   |                   ^^^^ `Send` is already required by `Trait`
   |
note: supertrait bound declared here
  --> src/main.rs:1:13
   |
1  | trait Trait: Send {}
   |              ^^^^
   = help: remove this bound
```

This is a subset of https://github.com/rust-lang/rust-clippy/issues/1463

**Comment by y21** [comment]
The lint [`implied_bounds_in_impls`](https://rust-lang.github.io/rust-clippy/master/index.html?search=implied#implied_bounds_in_impls) actually catches this exact pattern but only for trait bounds in `impls`, that is, this gets a warning today:
```rs
trait Trait: Send {}
fn foo(_: impl Trait + Send) {}
```
... so it might make sense to just extend that lint or base it off of its code, because the lint already has 99% of the necessary code for detecting this. I even left a comment in the code saying that we could easily lint trait bounds in where clauses, but the lint name doesn't make much sense anymore then.
https://github.com/rust-lang/rust-clippy/blob/65c339e71158fe15fc10d133a3d4decc076e4ef8/clippy_lints/src/implied_bounds_in_impls.rs#L330-L332

**Comment by oriongonza** [comment]
lol I should enable clippy in my editor before filing clippy issues, silly me
I love that the lint already exists, ty for it
